File: Medium/1921. Eliminate Maximum Number of Monsters/1921. Eliminate Maximum Number of Monsters.ts
function eliminateMaximum(dist: number[], speed: number[]): number {
    const n = dist.length;
    const timeToReach: number[] = [];

    for (let i = 0; i < n; i++) {
        timeToReach.push(Math.ceil(dist[i] / speed[i]));
    }

    timeToReach.sort((a, b) => a - b);

    let eliminated = 0;
    let time = 0;

    for (let i = 0; i < n; i++) {
        if (timeToReach[i] > time) {
            eliminated += 1;
            time += 1;
        } else {
            break;
        }
    }

    return eliminated;
}

File: Medium/1921. Eliminate Maximum Number of Monsters/1921. Eliminate Maximum Number of Monsters.rs
impl Solution {
    pub fn eliminate_maximum(dist: Vec<i32>, speed: Vec<i32>) -> i32 {
        let n = dist.len();
        let mut time_to_reach = vec![0; n];

        for i in 0..n {
            time_to_reach[i] = (dist[i] + speed[i] - 1) / speed[i];
        }

        time_to_reach.sort();

        let mut eliminated = 0;
        let mut time = 0;

        for i in 0..n {
            if time_to_reach[i] > time {
                eliminated += 1;
                time += 1;
            } else {
                break;
            }
        }

        eliminated
    }
}
